File: src/bgbg/engine/hittest.py
"""Hit-testing against the per-pixel lookup maps. Stdlib only.

The compute side emits three maps alongside the per-object masks:

  ``label``    the most *specific* object at each pixel (smallest painted last)
  ``general``  the most *general* object at each pixel (largest painted last)
  ``depth``    how many objects overlap at each pixel

Object ids are packed into an RGB image as ``R + G*256`` (0 = background), so up
to 65535 objects. ``depth`` is a single-channel count.

The engine does not decode PNGs — that is the shell's job (GdkPixbuf on the
desktop, an ImageBitmap/ImageData on the web). The shell hands over a raw pixel
buffer and its geometry; the engine just indexes it. That keeps this file free of
any image library and makes it trivially mirrorable in TypeScript over a
``Uint8Array``.
"""


class PixelMap:
    """A raw, shell-provided pixel buffer the engine can index directly.

    `data` is any bytes-like supporting integer indexing (bytes, memoryview,
    Uint8Array in the TS mirror). `channels` is the interleave, `stride` the row
    length in bytes.

    Each map carries its OWN width/height. The pre-extraction code bounds-checked
    the general and depth maps against the *label* map's dimensions, which is
    only correct because all three happen to be emitted at full image size.
    """

    __slots__ = ("data", "stride", "channels", "width", "height")

    def __init__(self, data, stride, channels, width, height):
        self.data = data
        self.stride = int(stride)
        self.channels = int(channels)
        self.width = int(width)
        self.height = int(height)

    def contains(self, x, y):
        return 0 <= x < self.width and 0 <= y < self.height

    def _offset(self, x, y):
        return y * self.stride + x * self.channels

    def value_at(self, x, y):
        """First channel (R, or the grey level of an L image loaded as RGB)."""
        if not self.contains(x, y):
            return 0
        return self.data[self._offset(x, y)]

    def id_at(self, x, y):
        """Unpack an object id: ``R + G*256``. 0 means background."""
        if not self.contains(x, y):
            return 0
        i = self._offset(x, y)
        if self.channels >= 2:
            return self.data[i] + (self.data[i + 1] << 8)
        return self.data[i]


class Hit:
    """What is under a point, in one shot."""

    __slots__ = ("specific", "general", "depth")

    def __init__(self, specific=0, general=0, depth=0):
        self.specific = specific
        self.general = general
        self.depth = depth

    @property
    def stacked(self):
        """True when several objects overlap here (whole vs part is meaningful)."""
        return self.depth >= 2 and self.general != self.specific

    def __repr__(self):
        return "Hit(specific=%d, general=%d, depth=%d)" % (
            self.specific, self.general, self.depth)


class HitMaps:
    """The three lookup maps for one segmented image."""

    __slots__ = ("label", "general", "depth")

    def __init__(self, label=None, general=None, depth=None):
        self.label = label          # PixelMap | None
        self.general = general      # PixelMap | None
        self.depth = depth          # PixelMap | None

    def clear(self):
        self.label = self.general = self.depth = None

    @property
    def loaded(self):
        return self.label is not None

    def specific_at(self, x, y):
        return self.label.id_at(x, y) if self.label is not None else 0

    def general_at(self, x, y):
        return self.general.id_at(x, y) if self.general is not None else 0

    def depth_at(self, x, y):
        return self.depth.value_at(x, y) if self.depth is not None else 0

    def hit(self, x, y):
        return Hit(self.specific_at(x, y),
                   self.general_at(x, y),
                   self.depth_at(x, y))

File: web/core/hittest.ts
/**
 * Hit-testing against the per-pixel lookup maps — the mirror of engine/hittest.py.
 *
 * The engine does not decode PNGs. The shell hands over a raw pixel buffer and
 * its geometry (GdkPixbuf on the desktop, ImageData/ImageBitmap here) and this
 * just indexes it — which is why the same code works over a Uint8Array.
 *
 * Object ids are packed as R + G*256. `depth` is a single-channel count, NOT an
 * id: decoding it as one gives 257 for a depth of one.
 */
export class PixelMap {
  constructor(
    public data: Uint8Array | number[],
    public stride: number,
    public channels: number,
    public width: number,
    public height: number,
  ) {}

  contains(x: number, y: number): boolean {
    return x >= 0 && x < this.width && y >= 0 && y < this.height;
  }

  private offset(x: number, y: number): number {
    return y * this.stride + x * this.channels;
  }

  /** First channel (R, or the grey level of an L image). */
  valueAt(x: number, y: number): number {
    if (!this.contains(x, y)) return 0;
    return this.data[this.offset(x, y)];
  }

  /** Unpack an object id: R + G*256. 0 means background. */
  idAt(x: number, y: number): number {
    if (!this.contains(x, y)) return 0;
    const i = this.offset(x, y);
    if (this.channels >= 2) return this.data[i] + (this.data[i + 1] << 8);
    return this.data[i];
  }
}

export interface Hit {
  specific: number;
  general: number;
  depth: number;
}

export class HitMaps {
  constructor(
    public label: PixelMap | null = null,
    public general: PixelMap | null = null,
    public depth: PixelMap | null = null,
  ) {}

  clear() {
    this.label = this.general = this.depth = null;
  }

  get loaded(): boolean {
    return this.label !== null;
  }

  specificAt(x: number, y: number): number {
    return this.label ? this.label.idAt(x, y) : 0;
  }

  generalAt(x: number, y: number): number {
    return this.general ? this.general.idAt(x, y) : 0;
  }

  depthAt(x: number, y: number): number {
    return this.depth ? this.depth.valueAt(x, y) : 0;
  }

  hit(x: number, y: number): Hit {
    return {
      specific: this.specificAt(x, y),
      general: this.generalAt(x, y),
      depth: this.depthAt(x, y),
    };
  }
}
